File: go/ff1_nist_test.go
package dealcode

import (
	"encoding/hex"
	"encoding/json"
	"os"
	"reflect"
	"strings"
	"testing"
)

// nistAlphabet maps character index to numeral value for the NIST sample
// vectors' plaintext/ciphertext strings.
const nistAlphabet = "0123456789abcdefghijklmnopqrstuvwxyz"

type nistVector struct {
	Sample     int    `json:"sample"`
	Cipher     string `json:"cipher"`
	KeyHex     string `json:"key_hex"`
	Radix      int    `json:"radix"`
	TweakHex   string `json:"tweak_hex"`
	Plaintext  string `json:"plaintext"`
	Ciphertext string `json:"ciphertext"`
}

func nistNumerals(t *testing.T, s string) []int {
	t.Helper()
	out := make([]int, len(s))
	for i := 0; i < len(s); i++ {
		idx := strings.IndexByte(nistAlphabet, s[i])
		if idx < 0 {
			t.Fatalf("character %q not in NIST alphabet", s[i])
		}
		out[i] = idx
	}
	return out
}

func TestFF1NISTVectors(t *testing.T) {
	raw, err := os.ReadFile("../testvectors/ff1_nist.json")
	if err != nil {
		t.Fatalf("read vectors: %v", err)
	}
	var file struct {
		Vectors []nistVector `json:"vectors"`
	}
	if err := json.Unmarshal(raw, &file); err != nil {
		t.Fatalf("parse vectors: %v", err)
	}
	if len(file.Vectors) != 9 {
		t.Fatalf("expected 9 NIST vectors, got %d", len(file.Vectors))
	}

	for _, vec := range file.Vectors {
		vec := vec
		t.Run(vec.Cipher, func(t *testing.T) {
			key, err := hex.DecodeString(vec.KeyHex)
			if err != nil {
				t.Fatalf("key_hex: %v", err)
			}
			tweak, err := hex.DecodeString(vec.TweakHex)
			if err != nil {
				t.Fatalf("tweak_hex: %v", err)
			}
			f, err := newFF1(key, vec.Radix)
			if err != nil {
				t.Fatalf("newFF1: %v", err)
			}
			pt := nistNumerals(t, vec.Plaintext)
			ct := nistNumerals(t, vec.Ciphertext)
			params, err := f.params(tweak, len(pt))
			if err != nil {
				t.Fatalf("params: %v", err)
			}

			if got := f.encrypt(params, pt); !reflect.DeepEqual(got, ct) {
				t.Errorf("sample %d encrypt: got %v, want %v", vec.Sample, got, ct)
			}
			if got := f.decrypt(params, ct); !reflect.DeepEqual(got, pt) {
				t.Errorf("sample %d decrypt: got %v, want %v", vec.Sample, got, pt)
			}
		})
	}
}

File: js/test/ff1-nist.test.mjs
// The 9 official NIST FF1-AES sample vectors (encrypt AND decrypt).
import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";

import { FF1 } from "../dist/esm/ff1.js";

const vectorsPath = fileURLToPath(
  new URL("../../testvectors/ff1_nist.json", import.meta.url),
);
const { vectors } = JSON.parse(readFileSync(vectorsPath, "utf8"));

// Character index = numeral value in the vector file.
const CHARS = "0123456789abcdefghijklmnopqrstuvwxyz";
const toNumerals = (s) => [...s].map((c) => CHARS.indexOf(c));
const toString = (xs) => xs.map((x) => CHARS[x]).join("");

assert.equal(vectors.length, 9);

for (const v of vectors) {
  test(`NIST FF1 sample ${v.sample} (${v.cipher}, radix ${v.radix})`, () => {
    const ff1 = new FF1(Buffer.from(v.key_hex, "hex"), v.radix);
    const tweak = Buffer.from(v.tweak_hex, "hex");
    const plain = toNumerals(v.plaintext);
    const cipher = toNumerals(v.ciphertext);

    assert.equal(toString(ff1.encrypt(tweak, plain)), v.ciphertext);
    assert.equal(toString(ff1.decrypt(tweak, cipher)), v.plaintext);
  });
}
